src/main/scala/viper/gobra/translator/implementations/components/OptionImpl.scala

```diff
@@ -0,0 +1,225 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) 2011-2020 ETH Zurich.
+
+package viper.gobra.translator.implementations.components
+
+import viper.gobra.translator.interfaces.Collector
+import viper.gobra.translator.interfaces.components.Options
+import viper.silver.{ast => vpr}
+
+class OptionImpl extends Options {
+  private val domainName : String = "Option"
+  private val typeVar : vpr.TypeVar = vpr.TypeVar("T")
+
+  /**
+    * Determines whether the Viper domain
+    * should be generated upon finalisation.
+    */
+  private var generateDomain : Boolean = false
+
+  /**
+    * Definition of the "optsome" function for the Viper domain of options:
+    *
+    * {{{
+    * function optsome(e : T) : Option[T]
+    * }}}
+    */
+  private lazy val optsome_func : vpr.DomainFunc = vpr.DomainFunc(
+    "optsome",
+    Seq(vpr.LocalVarDecl("e", typeVar)()),
+    vpr.DomainType(domainName, Map[vpr.TypeVar, vpr.Type]())(Seq(typeVar))
+  )(domainName = domainName)
+
+  /**
+    * Definition of the "optnone" function for the Viper domain of options:
+    *
+    * {{{
+    * function optnone() : Option[T]
+    * }}}
+    */
+  private lazy val optnone_func : vpr.DomainFunc = vpr.DomainFunc(
+    "optnone",
+    Seq(),
+    vpr.DomainType(domainName, Map[vpr.TypeVar, vpr.Type]())(Seq(typeVar))
+  )(domainName = domainName)
+
+  /**
+    * Definition of the "optget" function for the Viper domain of options:
+    *
+    * {{{
+    * function optget(o : Option[T]) : T
+    * }}}
+    */
+  private lazy val optget_func : vpr.DomainFunc = vpr.DomainFunc(
+    "optget",
+    Seq(vpr.LocalVarDecl("o", vpr.DomainType(domainName, Map[vpr.TypeVar, vpr.Type]())(Seq(typeVar)))()),
+    typeVar
+  )(domainName = domainName)
+
+  /**
+    * Definition of the "option_ex" axiom for the Viper domain of options:
+    *
+    * {{{
+    * axiom option_ex {
+    *   forall o : Option[T] :: o == optnone() || exists e : T :: o == optsome(e)
+    * }
+    * }}}
+    */
+  private lazy val option_ex_axiom : vpr.DomainAxiom = {
+    val oDecl = vpr.LocalVarDecl("o", vpr.DomainType(domainName, Map[vpr.TypeVar, vpr.Type]())(Seq(typeVar)))()
+    val eDecl = vpr.LocalVarDecl("e", typeVar)()
+
+    vpr.NamedDomainAxiom(
+      name = "option_ex",
+      exp = vpr.Forall(
+        Seq(oDecl),
+        Seq(),
```

Comment:
Please, add a trigger.

Reply:
Do you have a suggestion for a trigger for
```
axiom {
  forall o : Option[T] :: o == optnone() || exists e : T :: o == optsome(e)
}
```
or otherwise an alternative axiom?

Reply:
Nevermind, this axiom is superfluous